### PR TITLE
Rewrite editor-preview package as extensions of editor classes

### LINKED ISSUE #3634: [ui] Rename a file when the file is opened in Preview mode.
When having a multi-selection in the file navigator and the user selects "Rename" from the context menu, it generates an error:

```
root ERROR Error: Reference widget is not in the layout.
    at DockLayout.../../node_modules/@phosphor/widgets/lib/docklayout.js.DockLayout.addWidget (http://localhost:3000/bundle.js:11879:19)
    at TheiaDockPanel.../../node_modules/@phosphor/widgets/lib/dockpanel.js.DockPanel.addWidget (http://localhost:3000/bundle.js:13441:25)
    at TheiaDockPanel.../../packages/core/lib/browser/shell/theia-dock-panel.js.TheiaDockPanel.addWidget (http://localhost:3000/bundle.js:87969:36)
    at ApplicationShell.../../packages/core/lib/browser/shell/application-shell.js.ApplicationShell.addWidget (http://localhost:3000/bundle.js:85205:32)
    at FileSystemFrontendContribution.<anonymous> (http://localhost:3000/42.bundle.js:255:36)
    at step (http://localhost:3000/42.bundle.js:135:23)
    at Object.next (http://localhost:3000/42.bundle.js:116:53)
    at fulfilled (http://localhost:3000/42.bundle.js:107:58)
```

### LINKED ISSUE #4061: navigation from the editor preview widget does not work
For example from reference peek widget.

errors from console:
```
winjs.base.js:2057 Uncaught Error: Reference widget is not in the layout.
    at DockLayout.../../node_modules/@phosphor/widgets/lib/docklayout.js.DockLayout.addWidget (docklayout.js:323)
    at TheiaDockPanel.../../node_modules/@phosphor/widgets/lib/dockpanel.js.DockPanel.addWidget (dockpanel.js:303)
    at TheiaDockPanel.../../packages/core/lib/browser/shell/theia-dock-panel.js.TheiaDockPanel.addWidget (theia-dock-panel.ts:73)
    at ApplicationShell.../../packages/core/lib/browser/shell/application-shell.js.ApplicationShell.addWidget (application-shell.ts:668)
    at EditorManager.<anonymous> (widget-open-handler.ts:91)
    at step (widget-open-handler.ts:15)
    at Object.next (widget-open-handler.ts:15)
    at widget-open-handler.ts:15
    at new Promise (<anonymous>)
    at ../../packages/core/lib/browser/widget-open-handler.js.__awaiter (widget-open-handler.ts:15)
```

### LINKED ISSUE #4708: TypeError: Cannot read property 'addWidget' of null: cannot open editor widgets
I do not know how this happened but I have the snapshot link:
[Open in Gitpod](https://gitpod.io#snapshot/f4ab76db-4948-40ed-a084-b5c15577d417)

```
Uncaught (in promise) TypeError: Cannot read property 'addWidget' of null
    at EditorPreviewWidget.push.../../packages/editor-preview/lib/browser/editor-preview-widget.js.EditorPreviewWidget.attachPreviewWidget (editor-preview-widget.ts:113)
    at EditorPreviewWidget.push.../../packages/editor-preview/lib/browser/editor-preview-widget.js.EditorPreviewWidget.replaceEditorWidget (editor-preview-widget.ts:101)
    at EditorPreviewManager.<anonymous> (editor-preview-manager.ts:126)
    at step (editor-preview-manager.ts:15)
    at Object.next (editor-preview-manager.ts:15)
    at fulfilled (editor-preview-manager.ts:15)
push.../../packages/editor-preview/lib/browser/editor-preview-widget.js.EditorPreviewWidget.attachPreviewWidget @ editor-preview-widget.ts:113
push.../../packages/editor-preview/lib/browser/editor-preview-widget.js.EditorPreviewWidget.replaceEditorWidget @ editor-preview-widget.ts:101
(anonymous) @ editor-preview-manager.ts:126
step @ editor-preview-manager.ts:15
(anonymous) @ editor-preview-manager.ts:15
fulfilled @ editor-preview-manager.ts:15
postMessage (async)
registerImmediate @ setImmediate.js:120
setImmediate @ setImmediate.js:27
triggerMessageQueue @ main.js:240
callback @ main.js:289
../../node_modules/vscode-ws-jsonrpc/lib/socket/reader.js.WebSocketMessageReader.readMessage @ reader.ts:57
(anonymous) @ reader.ts:18
../../packages/core/lib/common/messaging/web-socket-channel.js.WebSocketChannel.handleMessage @ web-socket-channel.ts:49
WebSocketConnectionProvider.socket.onmessage @ ws-connection-provider.ts:58
```

### LINKED ISSUE #4903: alt+w doesn't close preview window
### Description
image

### Reproduction Steps
click file in explorer
press alt+w doesn't close preview window

**OS and Theia version:**
**Diagnostics:**

### LINKED ISSUE #6784: [navigator] race condition between click and dblClick handlers
### Description
When opening a file with a double click, both [handleClickEvent](https://github.com/eclipse-theia/theia/blob/25d0e2b3d38bb80e875cda93974ba6a2a7b9015d/packages/navigator/src/browser/navigator-widget.tsx#L233). and [handleDblClickEvent](https://github.com/eclipse-theia/theia/blob/25d0e2b3d38bb80e875cda93974ba6a2a7b9015d/packages/core/src/browser/tree/tree-widget.tsx#L1063) on navigator-widget.tsx are called.
The result is [open](https://github.com/eclipse-theia/theia/blob/25d0e2b3d38bb80e875cda93974ba6a2a7b9015d/packages/core/src/browser/opener-service.ts#L80) method inside opener service is called twice.

When working with editor-preview extension and `Editor.enablePreview` is true The following issues occur:
- Open any file with double click results in flickering behavior.
- Open a file with a single click, then open a second file with double click: in some environments the new file content is not rendered:
image

This issue happens because of a race condition in [open method of editor-preview-manager](https://github.com/eclipse-theia/theia/blob/25d0e2b3d38bb80e875cda93974ba6a2a7b9015d/packages/editor-preview/src/browser/editor-preview-manager.ts#L111) which is triggered twice, both from click handler and doubleClick handler.

### LINKED ISSUE #7078: preview: 'open preview' does not work for editor-previews
### Description
 Preview view not showing data in electron

### Reproduction Steps

Open a README.md file in electron
Open The same README.md with the preview view in electron
--> The data is not showing in the preview view

Open the same README.md file from VSCode with the preview view
--> The data shows in the preview editor view

**OS and Theia version:**
OS: Ubuntu 18.04
Theia: latest commit cc0da4e69f8141122496c119bb72f4df46584e76

**Diagnostics:** 

Theia electron:
ElectronNotShowingPreview

VSCode preview view:
VSCodeShowPreview

### LINKED ISSUE #8511: Regression when opening an .MD file in editor: I can see two tabs for one editor
See the screencast:

screencast 2020-09-16 16-49-47

- Theia Version: 993b7a8bc

### LINKED ISSUE #9485: Race condition in Editor Preview Manager and Editor Manager
### Bug Description:

If two open requests come in nearly simultaneously, one asking for a preview widget and one asking for a non-preview widget, a race condition in the opening routines of the `EditorPreviewManager` and `EditorManager` can lead to an empty editor preview next to an open normal editor and a broken shell / widget tracking system.

1. Install the [`vscode-clangd` plugin](https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.vscode-clangd) 
2. Find an `#include` directive pointing to an unambiguous filename in your workspace.
3. `Ctrl+Click` that filename.
4. You may (or may not) get bad behavior.

#### Alternatively:
1. Add a command like this somewhere in the repo:
```
        commands.registerCommand({ id: 'bad-opener-behavior', label: 'Try to open an editor two ways.' }, {
            execute: () => {
                const uri = new URI('<path to Theia>/devfile.yaml');
                open(this.openerService, uri, { preview: true });
                open(this.openerService, uri);
            }
        });
```
2. Run that command
3. You may (or may not) see something like this:

image

What appears to be happening is that the `EditorPreviewManager` checks for the existence of an editor for the file its trying to open (line 124), and if it doesn't find it, it does some other `async` work (lines 137ff) before trying to open a new editor (line 139 or 146) for that file. If, between the failed check and the actual opening of a new editor, the `EditorManager` has separately handled a request to open the same file, the two openers can come into conflict.

https://github.com/eclipse-theia/theia/blob/d734315e505ad4a7691a3f6f4a07d2209af0d9cf/packages/editor-preview/src/browser/editor-preview-manager.ts#L113-L143

I've only consistently reproduced the behavior on branches with #9369, but I don't believe that the bug is actually related to that change. 

### Additional Information

- Operating System: RHEL7
- Theia Version: 1.13.0

### PR DESCRIPTION
**Pull request by colin-grant-work**
#### What it does

This PR is a more-or-less wholesale rewrite of the `editor-preview` package to extend and override the functionality of the `EditorManager`, `EditorWidgetFactory`, and `EditorWidget` rather than using a wrapper around an `EditorWidget`.

 - Fixes #3634 by making the preview an immediate child of the shell
 - Fixes #4061 by using a normal editor widget
 - Fixes #4708 by reducing the complexity of the docking and widget-tracking logic
 - Fixes #6784, fixes #8511, fixes #9485 by implementing better checks for a widget's prior existence and by using a single opening routine for all editors, rather than having two competing openers for previews and regular editor widgets.

These issues aren't directly addressed by these changes, but I can't reproduce them with this code:
 - Closes #4903
 - Closes #7078

#### Regressions:
- ~~double-clicking the title of a preview editor will not pin that editor.~~
> I think the best way to handle this problem is what I've outlined in #9511, but in the meantime, I've achieved this by using a listener on the `document`.

#### Breaking Changes:
- ~~Change of various return types in `WidgetOpenHandler` and derived classes from `Promise<T | undefined>` to `MaybePromise<T> | undefined` to facilitate synchronous checks for widget existence.~~
> I've removed this breakage by adding methods to the `WidgetManager` and `WidgetOpenHandler` that can check for a widget's (pending) existence synchronously rather than modifying existing methods.

##### Basics
1. Makes sure that the preference `editor.enablePreview` is `true` and run the `Restore workbench layout` command.
2. Single-click various files in the navigator.
3. [ ] One editor should be created, and it should be replaced with each new file opened.
4. Double click the file you have open in the tree.
5. [ ] That editor should switch to non-preview mode.
6. Single click another file.
7. [ ] A new editor preview should open up.
8. Single click the file in the pinned (non-preview) editor.
9. [ ] The non-preview editor should be revealed, and no new editor should be opened.
10. Focus the current preview editor and start typing.
11. [ ] It should turn into a normal editor.

##### Fixes
 - Renaming
   1. Open a preview, then rename the file.
   2. [ ] The preview with the old name should be replaced with a new one with the new name.
   3. [ ] Single-clicking a different file should replace the preview.
 - Navigating
   1. Open an editor preview in a source code file.
   2. Use context menu 'Go to references' or <kbd>Shift + F12</kbd> to view a references peek widget
   3. [ ] Clicking on the destinations should work fine.
 - Dragging
   1. Open a preview.
   2. Drag it into a side panel
   3. [ ] You should still see the editor.
   4. Drag it back
   5. [ ] It should be a non-preview
 - Rapid opening
   1. Follow the reproduction instructions in #9485 
   2. [ ] You should never wind up with two editors.

##### Others
 - [ ] Using <kbd>ctrl+w</kbd> (Electron) / <kbd>alt+w</kbd> (Browser) to close an editor preview should work.
 - [ ] Opening a preview editor for e.g. a Markdown file and then using the `Open preview` should work as expected.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: Colin Grant <colin.grant@ericsson.com>

### CODE REVIEW

**Review by kenneth-marut-work — COMMENTED** [review]
This is working much better and it seems to have fixed the issues mentioned above. On a first pass running through the functionality, I only noticed two small behavioral issues related to pinning:

1. Splitting the editor view in the main area
- Open 2 files (one being a preview)
- Drag the preview to the right side of the MAIN area to split the screen
- The preview does not pin

2. Pinning an editor when it does not exist in the file tree
- Open a file using the steps in #9485 with a file that does not exist in your current workspace
- Observe that the file is opened as a preview widget, but it is not possible to pin it easily (now that the double clicking does not pin)
- In VSCode there is a context command for "Keep Editor Open" which might be a useful approach

File: packages/editor-preview/src/browser/editor-preview-manager.ts [hunk 1]
```diff
@@ -14,151 +14,104 @@
...
+export class EditorPreviewManager extends EditorManager {
     readonly id = EditorPreviewWidgetFactory.ID;
 
-    readonly label = 'Code Editor Preview';
-
-    protected currentEditorPreview: Promise<EditorPreviewWidget | undefined>;
-
-    @inject(EditorManager)
-    protected readonly editorManager: EditorManager;
+    @inject(WidgetManager) protected readonly widgetManager: WidgetManager;
```
Comment (kenneth-marut-work) [inline packages/editor-preview/src/browser/editor-preview-manager.ts:35]: Minor: This exists on the WidgetOpenHandler already

Comment (colin-grant-work) [reply packages/editor-preview/src/browser/editor-preview-manager.ts]: Good point - I've deleted it here.

**Comment by colin-grant-work** [comment]
@kenneth-marut-work, thanks for taking a look. I've made some changes to address your comments
 - changed the drag logic to track the widgets tabbar rather than its parent to cover dragging inside the `main` area.
 - added context menu commands and a keyboard shortcut parallel to VSCode's 'Keep Open' command.
 - brought back the pin-on-double-click logic with a listener on the `document`.

File: packages/editor-preview/src/browser/editor-preview-widget.ts [hunk 2]
```diff
@@ -15,22 +15,34 @@
...
     protected readonly onDidChangePreviewStateEmitter = new Emitter<void>();
     readonly onDidChangePreviewState = this.onDidChangePreviewStateEmitter.event;
 
+    protected readonly toDisposeOnLocationChange = new DisposableCollection();
+
     get isPreview(): boolean {
         return this._isPreview;
     }
 
+    constructor(
```
Comment (kenneth-marut-work) [inline packages/editor-preview/src/browser/editor-preview-widget.ts:37]: Might be an appropriate time to make this and the EditorWidget injectable

**Comment by kenneth-marut-work** [comment]
Verified that the new context menu pin option and double clicking work 👍

File: packages/core/src/browser/keybinding.ts [hunk 3]
```diff
@@ -245,6 +245,22 @@ export class KeybindingRegistry {
...
+        const getNumberOfKeystrokes = (binding: common.Keybinding): number => (binding.keybinding.trim().match(/\s/)?.length ?? 0) + 1;
+        const numberOfKeystrokesInBinding = getNumberOfKeystrokes(item);
+        const indexOfFirstItemWithEqualStrokes = scopedKeymap.findIndex(existingBinding => getNumberOfKeystrokes(existingBinding) === numberOfKeystrokesInBinding);
+        if (indexOfFirstItemWithEqualStrokes > -1) {
+            scopedKeymap.splice(indexOfFirstItemWithEqualStrokes, 0, item);
+        } else {
+            scopedKeymap.push(item);
+        }
+    }
+
```
Comment (colin-grant-work) [inline packages/core/src/browser/keybinding.ts:263]: This addition is necessary to ensure that this test passes:  
https://github.com/eclipse-theia/theia/blob/8ea255a4021393eba96a0da0acae3b1aa8f98ad6/examples/api-tests/src/keybindings.spec.js#L48-L63

Previously, it passed by coincidence - the relevant competition to the clear terminal command happened to be registered in an order that allowed the test to pass. 

FWIW, the assertion made in [this test](https://github.com/eclipse-theia/theia/blob/8ea255a4021393eba96a0da0acae3b1aa8f98ad6/examples/api-tests/src/keybindings.spec.js#L96-L115) that priority should depend on insertion order is pretty fragile, and seems designed to cover edge cases from the [`terminal` package](https://github.com/eclipse-theia/theia/blob/8ea255a4021393eba96a0da0acae3b1aa8f98ad6/packages/terminal/src/browser/terminal-frontend-contribution.ts#L430-L505) where we register a bunch of non-commands for <kbd>ctrl+\<key\></kbd> and <kbd>alt+\<key\></kbd> to get around bindings that already exist in the packages we use for terminals, but then *also* register some commands for specific combinations that we want to take priority over the dummy commands. There's got to be a better way to handle those requirements.

File: packages/editor-preview/src/browser/editor-preview-contribution.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import { ApplicationShell, KeybindingContribution, KeybindingRegistry, SHELL_TABBAR_CONTEXT_MENU, Widget } from '@theia/core/lib/browser';
+import { Command, CommandContribution, CommandRegistry, MenuContribution, MenuModelRegistry } from '@theia/core/lib/common';
+import { inject, injectable } from '@theia/core/shared/inversify';
+import { EditorPreviewWidget } from './editor-preview-widget';
+
+export namespace EditorPreviewCommands {
+    export const PIN_PREVIEW_COMMAND: Command = {
+        id: 'workbench.action.keepEditor',
+        label: 'Keep Open'
+    };
```
Comment (vince-fugnitto) [inline packages/editor-preview/src/browser/editor-preview-contribution.ts:26]:
```suggestion
    export const PIN_PREVIEW_COMMAND: Command = {
        id: 'workbench.action.keepEditor',
        category: 'View',
        label: 'Keep Editor'
    };
```

File: packages/editor-preview/src/browser/editor-preview-manager.ts [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /********************************************************************************
- * Copyright (C) 2018 Google and others.
+ * Copyright (C) 2021 Ericsson and others.
```
Comment (vince-fugnitto) [inline packages/editor-preview/src/browser/editor-preview-manager.ts:2]: We should keep the original license (generally we do not update them - they reflect the original author of the file).

File: packages/editor-preview/src/browser/editor-preview-widget-factory.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import URI from '@theia/core/lib/common/uri';
+import { EditorWidgetFactory } from '@theia/editor/lib/browser/editor-widget-factory';
+import { injectable } from '@theia/core/shared/inversify';
+import { EditorPreviewWidget } from './editor-preview-widget';
+import { NavigatableWidgetOptions } from '@theia/core/lib/browser';
+
+export interface EditorPreviewOptions extends NavigatableWidgetOptions {
```
Comment (vince-fugnitto) [inline packages/editor-preview/src/browser/editor-preview-widget-factory.ts:23]: Do we need both interfaces?
https://github.com/eclipse-theia/theia/pull/9518/files#diff-075aefa08a4fb3b0224a5bcc327bf6ed6f9b9aa1d8ff0fa9ad159b1a89406b3cR27

File: packages/editor-preview/src/browser/editor-preview-widget.ts [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /********************************************************************************
- * Copyright (C) 2018 Google and others.
+ * Copyright (C) 2021 Ericsson and others.
```
Comment (vince-fugnitto) [inline packages/editor-preview/src/browser/editor-preview-widget.ts:2]: Same as earlier comment.

File: packages/editor-preview/src/browser/editor-preview-manager.ts (hunk 5 above)
Comment (colin-grant-work) [reply packages/editor-preview/src/browser/editor-preview-manager.ts:2]: This is a slightly tricky case - I created whole new files with separate names (and new headers) while I was working on the code, then replaced the old files when I was done, so there is very little lineal connection between the file by this name on `master` and the version here - the diff shows almost only brackets 'carried over'. Since this is a totally breaking change (if you were importing anything from any of these files, it no longer does what you thought), there's also a case for using different file names so that's impossible to pretend that they still do what they used to. I'm not an expert on the ins and outs of the license headers, though, so I defer to the judgment of anyone better informed.

File: packages/editor-preview/src/browser/editor-preview-widget.ts (hunk 7 above)
Comment (colin-grant-work) [reply packages/editor-preview/src/browser/editor-preview-widget.ts:2]: changed to `2018-2021 Google and others`

File: packages/editor-preview/src/browser/editor-preview-contribution.ts (hunk 4 above)
Comment (colin-grant-work) [reply packages/editor-preview/src/browser/editor-preview-contribution.ts]: 👍 Done.

File: packages/editor-preview/src/browser/editor-preview-manager.ts (hunk 5 above)
Comment (colin-grant-work) [reply packages/editor-preview/src/browser/editor-preview-manager.ts]: Changed to `2018-2021 Google and others`

File: packages/editor-preview/src/browser/editor-preview-widget-factory.ts (hunk 6 above)
Comment (colin-grant-work) [reply packages/editor-preview/src/browser/editor-preview-widget-factory.ts:23]: Likely not. Kept the one in the factor, removed it here. 😅

**Review by vince-fugnitto — APPROVED** [review]
I verified the following functionality :+1: 

**Preference `editor.enablePreview`**:

- [x] the preference is enabled by default
- [x] when _true_ editors are opened as a preview and replaced
- [x] when _false_ editors are opened normally and not replaced
- [x] when toggling from _true_ to _false_, the open _preview_ editor is transformed into a normal editor 

**Application**

- [x] a _preview_ editor is transformed into a normal editor when it is modified by a user
- [x] a _single-click_ in the explorer opens a _preview_ editor when the preference is on
- [x] a _single-click_ in the explorer opens a normal editor when the preference is off
- [x] a double-click in the explorer opens a normal editor (preference on)
- [x] renaming a _preview_ editor works correctly - the editor remains a _preview_
- [x] dragging _preview_ editors works correctly (sidepanel to main-area and so-on)
- [x] keybinding successfully closes a _preview_ editor
- [x] splitting a _preview_ editor works
- [x] reloading the application with _preview_ editors works 
- [x] performing toolbar actions (ex: 'markdown: open preview') should work for _preview_ editors

**Comment by colin-grant-work** [comment]
@caseyflynn-google, this may be of interest to you. I plan to merge on Monday or Tuesday, if there are no objections.

File: packages/editor-preview/src/browser/editor-preview-contribution.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+import { ApplicationShell, KeybindingContribution, KeybindingRegistry, SHELL_TABBAR_CONTEXT_MENU, Widget } from '@theia/core/lib/browser';
+import { Command, CommandContribution, CommandRegistry, MenuContribution, MenuModelRegistry } from '@theia/core/lib/common';
+import { inject, injectable } from '@theia/core/shared/inversify';
+import { EditorPreviewWidget } from './editor-preview-widget';
+
+export namespace EditorPreviewCommands {
+    export const PIN_PREVIEW_COMMAND: Command = {
+        id: 'workbench.action.keepEditor',
+        category: 'View',
+        label: 'Keep Editor',
```
Comment (kenneth-marut-work) [inline packages/editor-preview/src/browser/editor-preview-contribution.ts:26]: minor: in VSCode, this label that appears in the context menu is 'Keep Open'

Comment (vince-fugnitto) [reply packages/editor-preview/src/browser/editor-preview-contribution.ts:26]: > minor: in VSCode, this label that appears in the context menu is 'Keep Open'

Then we would need to update the label at: https://github.com/eclipse-theia/theia/pull/9518/files#diff-14f606c80287893ed697301e733999e7682fe47f8680bb99d05feeb40f446d16R64.

**Comment by kenneth-marut-work** [comment]
Found one small issue, right clicking a file and using the "Open With" menu option reveals 2 "Code Editor" openers

image

File: packages/editor-preview/src/browser/editor-preview-contribution.ts (hunk 8 above)
Comment (colin-grant-work) [reply packages/editor-preview/src/browser/editor-preview-contribution.ts:26]: I've modified the label below.

**Comment by colin-grant-work** [comment]
> Found one small issue, right clicking a file and using the "Open With" menu option reveals 2 "Code Editor" openers

@kenneth-marut-work, good find. Since `WidgetOpenHandler` was already bound to the `EditorManager`, I didn't need to bind it to the `EditorPreviewManager` again. I've removed that, and now there should be 1 (or 0) 'Open With' entries.

**Review by caseyflynn-google — APPROVED** [review]
Awesome work! Verified all items listed in the testing section:

> #### How to test
> ##### Basics
> 1. Makes sure that the preference `editor.enablePreview` is `true` and run the `Restore workbench layout` command.
> 2. Single-click various files in the navigator.
> 3. [x]  One editor should be created, and it should be replaced with each new file opened.
> 4. Double click the file you have open in the tree.
> 5. [x]  That editor should switch to non-preview mode.
> 6. Single click another file.
> 7. [x]  A new editor preview should open up.
> 8. Single click the file in the pinned (non-preview) editor.
> 9. [x]  The non-preview editor should be revealed, and no new editor should be opened.
> 10. Focus the current preview editor and start typing.
> 11. [x]  It should turn into a normal editor.
> 
> ##### Fixes
> * Renaming
>   
>   1. Open a preview, then rename the file.
>   2. [x]  The preview with the old name should be replaced with a new one with the new name.
>   3. [x]  Single-clicking a different file should replace the preview.
> * Navigating
>   
>   1. Open an editor preview in a source code file.
>   2. Use context menu 'Go to references' or Shift + F12 to view a references peek widget
>   3. [x]  Clicking on the destinations should work fine.
> * Dragging
>   
>   1. Open a preview.
>   2. Drag it into a side panel
>   3. [x]  You should still see the editor.
>   4. Drag it back
>   5. [x]  It should be a non-preview
> * Rapid opening
>   
>   1. Follow the reproduction instructions in #9485
>   2. [x]  You should never wind up with two editors.
> 
> ##### Others
> * [x]  Using ctrl+w (Electron) / alt+w (Browser) to close an editor preview should work.
> * [x]  Opening a preview editor for e.g. a Markdown file and then using the `Open preview` should work as expected.
> 

Additional checks:
- [x] All context menu items in editor preview tab work (Close, Close Others, Toggle Maximized etc)
- [x] Language server features work in editor preview
- [x] Editor reviews are properly revived (only tested in browser)
